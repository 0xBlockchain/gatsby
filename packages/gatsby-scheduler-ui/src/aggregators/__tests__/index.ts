import aggregator from "../index"

// Mock Date.now
Date.now = jest.fn(() => 1487076708000)

describe(`file updates`, () => {
  it(`sums the total bytes correctly`, () => {
    let root = { files: {}, hosts: {} }
    root = aggregator(root, {
      type: `fileChunkUpdate`,
      data: {
        url: `http://example.com/icon.png`,
        host: `example.com`,
        bytes: 2,
        time: 1,
      },
    })
    root = aggregator(root, {
      type: `fileChunkUpdate`,
      data: {
        url: `http://example.com/icon2.png`,
        host: `example.com`,
        bytes: 5,
        time: 2,
      },
    })
    root = aggregator(root, {
      type: `fileChunkUpdate`,
      data: {
        url: `http://example.com/icon.png`,
        host: `example.com`,
        bytes: 5,
        time: 3,
      },
    })
    root = aggregator(root, {
      type: `fileChunkUpdate`,
      data: {
        url: `http://example.com/icon2.png`,
        host: `example.com`,
        bytes: 1,
        time: 4,
      },
    })
    root = aggregator(root, {
      type: `fileChunkUpdate`,
      data: {
        url: `http://example2.com/icon2.png`,
        host: `example2.com`,
        bytes: 1,
        time: 4,
      },
    })

    expect(root).toMatchSnapshot()
    expect(Object.keys(root.files)).toHaveLength(3)
    expect(root.files[`http://example.com/icon2.png`].totalBytes).toBe(6)
    expect(root.hosts[`example.com`].totalBytes).toBe(13)
  })

  it(`updates stream speeds correctly`, () => {
    let root = { files: {}, hosts: {} }
    root = aggregator(root, {
      type: `streamSpeedUpdate`,
      data: {
        host: `example.com`,
        speed: 2,
      },
    })
    expect(root).toMatchSnapshot()
    expect(root.hosts[`example.com`].speed).toEqual(2)
    root = aggregator(root, {
      type: `streamSpeedUpdate`,
      data: {
        host: `example.com`,
        speed: 5,
      },
    })
    expect(root).toMatchSnapshot()
    expect(root.hosts[`example.com`].speed).toEqual(5)
  })

  it(`reduces worker roots correctly`, () => {
    let root1 = { files: {}, hosts: {} }
    let root2 = { files: {}, hosts: {} }
    let root3 = { files: {}, hosts: {} }

    // Update speeds
    root1 = aggregator(root1, {
      type: `streamSpeedUpdate`,
      data: {
        host: `example.com`,
        speed: 2,
      },
    })
    root2 = aggregator(root2, {
      type: `streamSpeedUpdate`,
      data: {
        host: `example.com`,
        speed: 3,
      },
    })

    // File downloading
    root1 = aggregator(root1, {
      type: `fileChunkUpdate`,
      data: {
        url: `http://example.com/icon.png`,
        host: `example.com`,
        bytes: 50,
        time: 1,
      },
    })
    root2 = aggregator(root2, {
      type: `fileChunkUpdate`,
      data: {
        url: `http://example.com/icon2.png`,
        host: `example.com`,
        bytes: 1,
        time: 2,
      },
    })

    let combinedRoot = {}
    combinedRoot = aggregator(combinedRoot, {
      type: `combineRoots`,
      data: [root1, root2, root3],
    })
    expect(combinedRoot).toMatchSnapshot()
    expect(combinedRoot.total.speed).toBe(5)
    expect(combinedRoot.total.totalBytes).toBe(51)
  })
})
