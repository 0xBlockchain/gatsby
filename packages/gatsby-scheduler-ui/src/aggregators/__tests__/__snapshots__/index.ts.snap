// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`file updates reduces worker roots correctly 1`] = `
Object {
  "end": 2,
  "example.com": Object {
    "speed": 5,
    "totalBytes": 51,
  },
  "hostNames": Array [
    "example.com",
    "total",
  ],
  "start": 1,
  "time": 1487076708,
  "total": Object {
    "speed": 5,
    "totalBytes": 51,
  },
}
`;

exports[`file updates sums the total bytes correctly 1`] = `
Object {
  "end": 4,
  "files": Object {
    "http://example.com/icon.png": Object {
      "host": "example.com",
      "totalBytes": 7,
      "url": "http://example.com/icon.png",
    },
    "http://example.com/icon2.png": Object {
      "host": "example.com",
      "totalBytes": 6,
      "url": "http://example.com/icon2.png",
    },
    "http://example2.com/icon2.png": Object {
      "host": "example2.com",
      "totalBytes": 1,
      "url": "http://example2.com/icon2.png",
    },
  },
  "hosts": Object {
    "example.com": Object {
      "speed": 0,
      "totalBytes": 13,
    },
    "example2.com": Object {
      "speed": 0,
      "totalBytes": 1,
    },
    "total": Object {
      "speed": 0,
      "totalBytes": 14,
    },
  },
  "start": 1,
}
`;

exports[`file updates updates stream speeds correctly 1`] = `
Object {
  "files": Object {},
  "hosts": Object {
    "example.com": Object {
      "speed": 2,
      "totalBytes": 0,
    },
    "total": Object {
      "speed": 2,
      "totalBytes": 0,
    },
  },
}
`;

exports[`file updates updates stream speeds correctly 2`] = `
Object {
  "files": Object {},
  "hosts": Object {
    "example.com": Object {
      "speed": 5,
      "totalBytes": 0,
    },
    "total": Object {
      "speed": 5,
      "totalBytes": 0,
    },
  },
}
`;
